### PR TITLE
Re-re-rewire data/form state up to redux

### PR DESCRIPTION
**Pull request by davidjoy**
This PR rewires how our forms get their data.  It also folds in changes from #35 because @abutterworth and I had a pretty hefty conflict between our two branches.

This turned into a broader effort than I intended, admittedly.  

The original goal was to take our form state out of component state and put it into redux state for all six forms on the profile page.  This has been done, but doing so begged a bit of broader refactoring and renaming as well.

In no particular order:

1. We were referring to our forms as fields - that language has been fixed.  They're forms now.
2. We were putting 'draft' form values in component state - that state has been moved up to the redux level.
3. The fetchProfile action is now responsible for making several underlying calls to the system for the profile's account information, preferences (if it's your own account), and certificates.  Before, we had multiple actions responsible for each part, and coordination was hard.
4. `react-router-redux` has been replaced with `connected-react-router`.  This is admittedly not something that ultimately needed to be in this PR - I _thought_ I was going to need it, but after the work was done, the need for the change fell out and it's essentially unrelated.  That said, `react-router-redux` is deprecated and no longer supported, and `connected-react-router` has taken its place.  I didn't see any reason to throw away the work, so here it is.
5. I updated some packages as part of number 4 above which didn't strictly need to be updated.  I _thought_ I needed to update them w/r/t changing the router, but I didn't.  As above, didn't see any reason to throw away the work, though.
6. Introduced `reselect` to handle derived state.  This helps keep components clean. 
7. Directly connected the forms to the store.  This makes ProfilePage.jsx a lot cleaner.
8. The ProfilePage.jsx file still manages calling action creators to put stuff in redux - the forms are connected for their data, but are otherwise "dumb" about how the data gets back into redux.  It felt weird, for instance, to have the Name.jsx component call an action creator called `saveProfile`... just seemed above its pay grade.  Admittedly this is a bit asymmetrical, but it allowed ProfilePage.jsx to be a lot shorter/have less responsibility for passing data down.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "bootstrap": "^4.2.1",
     "camelcase-keys": "^5.0.0",
     "classnames": "^2.2.5",
+    "connected-react-router": "^5.0.1",
```
Comment (davidjoy) [inline package.json]: As said in the summary, changing the router and upgrading the react/redux packages wasn't strictly necessary, but I didn't see any reason to waste the work.

File: src/actions/ProfileActions.js [hunk 2]
```diff
@@ -1,42 +1,31 @@
...
+  type: FETCH_PROFILE.BASE,
+  payload: { username },
 });
 
 export const fetchProfileBegin = () => ({
   type: FETCH_PROFILE.BEGIN,
 });
 
-export const fetchProfileSuccess = profile => ({
+export const fetchProfileSuccess = (account, preferences, certificates) => ({
```
Comment (davidjoy) [inline src/actions/ProfileActions.js]: fetchProfileSuccess dumps all our fancy new data into the reducer now - not just the misnomered "profile" (which is really "account" data)

File: src/actions/ProfileActions.js [hunk 3]
```diff
@@ -48,17 +37,25 @@ export const fetchProfileReset = () => ({
   type: FETCH_PROFILE.RESET,
 });
 
-export const fetchProfile = username => ({
-  type: FETCH_PROFILE.BASE,
-  payload: { username },
+// SAVE PROFILE ACTIONS
+
+export const saveProfile = formId => ({
```
Comment (davidjoy) [inline src/actions/ProfileActions.js]: To save a profile, we just need to pass which form was being submitted.  The Saga can get the full data out of the reducer itself.

File: src/components/ProfilePage.jsx [hunk 4]
```diff
@@ -9,152 +9,78 @@ import {
...
 import SocialLinks from './ProfilePage/SocialLinks';
 import Bio from './ProfilePage/Bio';
-import MyCertificates from './ProfilePage/MyCertificates';
+import Certificates from './ProfilePage/Certificates';
 
 export class ProfilePage extends React.Component {
   constructor(props) {
     super(props);
 
-    this.state = {
```
Comment (davidjoy) [inline src/components/ProfilePage.jsx]: Goodnight, sweet state-prince.

File: src/components/ProfilePage.jsx [hunk 5]
```diff
@@ -9,152 +9,78 @@ import {
...
+  updateVisibilityDraft,
+  updateAccountDraft,
 } from '../actions/ProfileActions';
 
 // Components
 import ProfileAvatar from './ProfilePage/ProfileAvatar';
-import FullName from './ProfilePage/FullName';
-import UserLocation from './ProfilePage/UserLocation';
+import Name from './ProfilePage/Name';
+import Country from './ProfilePage/Country';
```
Comment (davidjoy) [inline src/components/ProfilePage.jsx]: I aligned some names with what the server sends.

File: src/components/ProfilePage.jsx [hunk 6]
```diff
@@ -9,152 +9,78 @@ import {
...
-
-  onVisibilityChange(fieldName, visibility) {
-    this.setState({
-      [fieldName]: {
-        value: this.state[fieldName].value,
-        visibility,
-      },
-    });
+  handleChange(formId, name, value) {
+    if (name === 'visibility') {
```
Comment (davidjoy) [inline src/components/ProfilePage.jsx]: Depending on what kind of field changed, we store the data in different places in the reducer.  Separate actions felt like a relatively clean way to tell the difference.

File: src/components/ProfilePage.jsx (hunk 4 above)
Comment (albemarle) [reply src/components/ProfilePage.jsx]: 🤴 💀 💤

File: src/sagas/RootSaga.js [hunk 7]
```diff
@@ -24,93 +24,75 @@ import {
...
+    }
+
+    const result = yield all(calls);
+
+    if (result.length > 2) {
+      const [account, certificates, preferences] = result;
+      yield put(fetchProfileSuccess(account, preferences, certificates));
+    } else {
+      const [account, certificates] = result;
+      yield put(fetchProfileSuccess(account, { visibility: {} }, certificates));
```
Comment (abutterworth) [inline src/sagas/RootSaga.js]: This may be a little weird, recognize the need tho

File: src/sagas/RootSaga.js [hunk 8]
```diff
@@ -24,93 +24,75 @@ import {
...
-      responseData.profile = yield call(
-        ProfileApiService.patchProfile,
-        username,
-        profileData,
-      );
+    let accountResult = null;
+    // Build the visibility drafts into a structure the API expects.
+    const preferences = {
+      visibility: visibilityDrafts,
+    };
```
Comment (abutterworth) [inline src/sagas/RootSaga.js]: Let's keep in mind that we need to afford for other preferences. For example:
```
{
   accountPrivacy: 'custom',
   visibility: { /* field visibilities */ },
}

File: src/selectors/ProfilePageSelector.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  (username, profileAccount) => username === profileAccount.username,
+);
+
+export const editableFormModeSelector = createSelector(
+  profileAccountSelector,
+  profileCertificatesSelector,
+  formIdSelector,
+  isCurrentUserProfileSelector,
+  currentlyEditingFieldSelector,
+  (account, certificates, formId, isCurrentUserProfile, currentlyEditingField) => {
```
Comment (abutterworth) [inline src/selectors/ProfilePageSelector.js]: FWIW I'm not a huge fan of this composable selector pattern

File: src/actions/ProfileActions.js [hunk 10]
```diff
@@ -1,42 +1,31 @@
...
-  type: FIELD_CLOSE,
-  fieldName,
-});
-
-export const updateDrafts = drafts => ({
-  type: UPDATE_DRAFTS,
-  drafts,
+export const fetchProfile = username => ({
+  type: FETCH_PROFILE.BASE,
+  payload: { username },
```
[the remark this answers is not in the document]
Comment (davidjoy) [reply src/actions/ProfileActions.js]: The payload thing is a "flux standard action" convention.  It's supposed to be the canonical way an action is structured.  That said, we don't seem to follow the spec completely w/r/t errors:

https://github.com/redux-utilities/flux-standard-action

File: src/sagas/RootSaga.js (hunk 7 above)
Comment (davidjoy) [reply src/sagas/RootSaga.js]: Yeah, I couldn't figure out a more elegant way to handle it.

File: src/selectors/ProfilePageSelector.js (hunk 9 above)
Comment (davidjoy) [reply src/selectors/ProfilePageSelector.js]: It's a bit weird to get used to, admittedly.  It really starts shining when you need lots of different kinds of data in lots of different areas of the app...

**Pull request merged**